File: course/string/O.cpp
#include <bits/stdc++.h>

using namespace std;

using ll = int64_t;
using ld = long double;
using ii = pair<int, int>;

const char SPACE = ' ', ENDL = '\n';

//////////////////////////////////////////////////////////////////

template <typename T>
T read() {
    T value;
    cin >> value;
    return value;
}

int ri() {
    return read<int>();
}

ll rll() {
    return read<ll>();
}

string rs() {
    return read<string>();
}

template<typename T>
vector<T> rv(int size) {
    vector<T> a(size);
    for (int i = 0; i < size; ++i) {
        a[i] = read<T>();
    }
    return a;
}

//////////////////////////////////////////////////////////////////

bool yn(bool result, string yes = "YES", string no = "NO") {
	cout << (result ? yes : no) << ENDL;
	return result;
}

//////////////////////////////////////////////////////////////////

vector<string> split(string& s, string seps) {
    vector<string> parts;

    string part = "";
    auto add_part = [&]() {
        parts.push_back(part);
        part = "";
    };

    for (char ch : s) {
        if (string::npos != seps.find(ch)) {
            add_part();
        } else {
            part += ch;
        }
    }

    add_part();

    return parts;
}

//////////////////////////////////////////////////////////////////

ld sqrt_ld(ld value) {
    return sqrt(value);
}

//////////////////////////////////////////////////////////////////

void init_io() {
    string file_name = "";
    if (!file_name.empty()) {
        string input_file_name = file_name + ".in";
        freopen(input_file_name.c_str(), "rt", stdin);

        string output_file_name = file_name + ".out";
        freopen(output_file_name.c_str(), "wt", stdout);
    } else {
        #ifndef ONLINE_JUDGE
        string input_file_name = "../../input.txt"; // reading from file in parent of parent of directory with file
        freopen(input_file_name.c_str(), "rt", stdin);
        #endif
    }

	ios::sync_with_stdio(false);
	cin.tie(nullptr);

	cout << setprecision(10) << fixed;
}

ii get_ans(string& s, string& t) {
    char start = 'a', end = 'z';
    int size = end - start + 1;

    vector<int> s_counts(size, 0);
    for (char ch : s) {
        s_counts[ch - start]++;
    }

    vector<int> t_counts(size, 0);
    for (char ch : t) {
        t_counts[ch - start]++;
    }

    ii need_tree = { false, false };

    bool need_remove = false;
    for (int ch = 0; ch < size; ++ch) {
        int delta = s_counts[ch] - t_counts[ch];

        if (delta < 0) return need_tree;

        if (delta > 0) {
            need_remove = true;
        }
    }

    int ti = 0;
    for (int si = 0; si < s.length() && ti < t.length(); ++si) {
        if (s[si] == t[ti]) {
            ++ti;
        }
    }

    bool need_swap = (ti != t.length());

    return { need_remove, need_swap } ;
}

void solve() {
    string s = rs();
    string t = rs();

    ii ans = get_ans(s, t);

    vector<string> ans_names = {
        "need tree", "automaton", "array", "both"
    };

    int name_index = (ans.first << 0) | (ans.second << 1);
    cout << ans_names[name_index] << ENDL;
}

int main() {
    init_io();
    solve();
    return 0;
}

File: course/string/O.java
import java.awt.Point;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.util.Locale;
import java.util.StringTokenizer;

public class O {

    private void solve() {
        char[] s = readString().toCharArray();
        char[] t = readString().toCharArray();

        Point ans = getAnswer(s, t);

        String[] names = {
                "need tree", "automaton", "array", "both"
        };

        int nameIndex = ans.x | (ans.y << 1);
        out.println(names[nameIndex]);
    }

    Point getAnswer(char[] s, char[] t) {
        char start = 'a', end = 'z';
        int size = end - start + 1;

        int[] sCounts = new int[size];
        for (char ch : s) sCounts[ch - start]++;

        int[] tCounts = new int[size];
        for (char ch : t) tCounts[ch - start]++;

        int notNeeded = 0, needed = 1;
        Point needTree = new Point(notNeeded, notNeeded);

        boolean needRemove = false;
        for (int ch = 0; ch < size; ++ch) {
            int delta = sCounts[ch] - tCounts[ch];
            if (delta < 0) return needTree;
            if (delta > 0) {
                needRemove = true;
            }
        }

        int ti = 0;
        for (int si = 0; si < s.length && ti < t.length; ++si) {
            if (s[si] == t[ti]) {
                ++ti;
            }
        }

        boolean needSwap = (ti != t.length);

        return new Point(
                needRemove ? needed : notNeeded,
                needSwap ? needed : notNeeded
        );
    }

    //////////////////////////////////////////////////////////////////

    long sqrtLong(long x) {
        long root = (long)Math.sqrt(x);
        while (root * root > x) --root;
        while ((root + 1) * (root + 1) <= x) ++root;
        return root;
    }

    //////////////////////////////////////////////////////////////////

    private boolean yesNo(boolean yes) {
        return yesNo(yes, "YES", "NO");
    }

    private boolean yesNo(boolean yes, String yesString, String noString) {
        out.println(yes ? yesString : noString);
        return yes;
    }

    //////////////////////////////////////////////////////////////////

    private long readLong() {
        return Long.parseLong(readString());
    }

    private int[] readIntArray(int n) {
        int[] a = new int[n];
        for (int i = 0; i < n; ++i) a[i] = readInt();
        return a;
    }

    private int readInt() {
        return Integer.parseInt(readString());
    }

    private String readString() {
        while (!tok.hasMoreTokens()) {
            String nextLine = readLine();
            if (null == nextLine) return null;
            tok = new StringTokenizer(nextLine);
        }

        return tok.nextToken();
    }

    private String readLine() {
        try {
            return in.readLine();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    //////////////////////////////////////////////////////////////////

    private BufferedReader in;
    private StringTokenizer tok;
    private PrintWriter out;

    private void initFileIO(String inputFileName, String outputFileName) throws FileNotFoundException {
        in = new BufferedReader(new FileReader(inputFileName));
        out = new PrintWriter(outputFileName);
    }

    private void initConsoleIO() throws IOException {
        in = new BufferedReader(new InputStreamReader(System.in));
        out = new PrintWriter(System.out);
    }

    private void initIO() throws IOException {
        Locale.setDefault(Locale.US);

        String fileName = "";

        if (!fileName.isEmpty()) {
            initFileIO(fileName + ".in", fileName + ".out");
        } else {
            if (new File("input.txt").exists()) {
                initFileIO("input.txt", "output.txt");
            } else {
                initConsoleIO();
            }
        }

        tok = new StringTokenizer("");
    }

    //////////////////////////////////////////////////////////////////

    private void run() {
        try {
            long timeStart = System.currentTimeMillis();

            initIO();
            solve();
            out.close();

            long timeEnd = System.currentTimeMillis();
            System.err.println("Time(ms) = " + (timeEnd - timeStart));
        } catch (Exception e) {
            e.printStackTrace();
            throw new RuntimeException(e);
        }
    }

    public static void main(String[] args) {
        new O().run();
    }
}
